File: py/eval.py
"""Prefix-notation evaluator built on top of parse.py tokenizer."""

from typing import Any, Callable, Dict, List, Optional, Tuple

from parse import TokenType, tokenize


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------

class Return(Exception):
    """Raised to unwind the stack on (return value)."""
    def __init__(self, value: Any) -> None:
        self.value = value


# ---------------------------------------------------------------------------
# AST node types
# ---------------------------------------------------------------------------

class Symbol:
    """A named reference (variable, operator, keyword)."""
    __slots__ = ("name",)
    def __init__(self, name: str) -> None:
        self.name = name

    def __repr__(self) -> str:
        return f"Symbol({self.name!r})"


class SExpr:
    """An S-expression list."""
    __slots__ = ("items",)
    def __init__(self, items: List[Any]) -> None:
        self.items = items

    def __repr__(self) -> str:
        return f"SExpr({self.items!r})"


class Lambda:
    """User-defined function with closure environment."""
    __slots__ = ("params", "body", "closure")
    def __init__(self, params: List[str], body: Any, closure: "Env") -> None:
        self.params = params
        self.body = body
        self.closure = closure

    def __repr__(self) -> str:
        return f"<lambda ({' '.join(self.params)})>"


class Builtin:
    """Built-in callable (binary or unary)."""
    __slots__ = ("func", "arity", "name")
    def __init__(self, func: Callable, arity: int, name: str = "") -> None:
        self.func = func
        self.arity = arity
        self.name = name

    def __repr__(self) -> str:
        return f"<builtin {self.name or '?'}/{self.arity}>"


# ---------------------------------------------------------------------------
# Environment (lexical scoping via parent chain)
# ---------------------------------------------------------------------------

class Env:
    """A scope with bindings and optional parent."""
    __slots__ = ("_bindings", "_parent")

    def __init__(self, parent: Optional["Env"] = None) -> None:
        self._bindings: Dict[str, Any] = {}
        self._parent = parent

    def define(self, name: str, value: Any) -> None:
        """Add binding to *this* scope (shadows parent)."""
        self._bindings[name] = value

    def lookup(self, name: str) -> Any:
        """Walk parent chain to find a binding."""
        if name in self._bindings:
            return self._bindings[name]
        if self._parent is not None:
            return self._parent.lookup(name)
        raise NameError(f"undefined: {name}")

    def set(self, name: str, value: Any) -> None:
        """Mutate existing binding (walks parent chain), or define in
        current scope if not found anywhere."""
        if name in self._bindings:
            self._bindings[name] = value
            return
        if self._parent is not None:
            try:
                self._parent.set(name, value)
                return
            except NameError:
                pass
        self._bindings[name] = value  # define locally

    def extend(self) -> "Env":
        """Create a child env (for function call, do block)."""
        return Env(parent=self)


# ---------------------------------------------------------------------------
# Parser: flat token list → nested AST
# ---------------------------------------------------------------------------

def _token_to_atom(tok: Tuple[TokenType, Any]) -> Any:
    """Convert a single token tuple into an AST atom."""
    kind, value = tok
    if kind == TokenType.WORD:
        return Symbol(value)
    return value  # literal: int, float, bool, None, str


def parse(tokens: List[Tuple[TokenType, Any]]) -> Any:
    """Recursive descent over flat tokens, returning a single AST node."""
    it = iter(enumerate(tokens))

    def _parse_list() -> SExpr:
        items: List[Any] = []
        for _idx, tok in it:
            kind, value = tok
            if kind == TokenType.WORD and value == "(":
                items.append(_parse_list())
            elif kind == TokenType.WORD and value == ")":
                return SExpr(items)
            else:
                items.append(_token_to_atom(tok))
        raise SyntaxError("unexpected EOF inside list")

    # Top-level: wrap in an implicit do-list if multiple top-level forms
    top: List[Any] = []
    for _idx, tok in it:
        kind, value = tok
        if kind == TokenType.WORD and value == "(":
            top.append(_parse_list())
        elif kind == TokenType.WORD and value == ")":
            raise SyntaxError("unexpected ')'")
        else:
            top.append(_token_to_atom(tok))

    if len(top) == 1:
        return top[0]
    return SExpr([Symbol("do")] + top)


# ---------------------------------------------------------------------------
# Built-in operators
# ---------------------------------------------------------------------------

def _make_builtins() -> Dict[str, Builtin]:
    b: Dict[str, Builtin] = {}

    def _bin(name: str, fn: Callable[[Any, Any], Any]) -> None:
        b[name] = Builtin(fn, 2, name)

    def _una(name: str, fn: Callable[[Any], Any]) -> None:
        b[name] = Builtin(fn, 1, name)

    # --- arithmetic ---
    _bin("+", lambda a, b: a + b)
    _bin("-", lambda a, b: a - b)
    _bin("*", lambda a, b: a * b)
    _bin("/", lambda a, b: a / b)
    _bin("mod", lambda a, b: a % b)

    # --- comparison ---
    _bin("<", lambda a, b: a < b)
    _bin(">", lambda a, b: a > b)
    _bin("<=", lambda a, b: a <= b)
    _bin(">=", lambda a, b: a >= b)
    _bin("=", lambda a, b: a == b)
    _bin("!=", lambda a, b: a != b)

    # --- logic ---
    _bin("and", lambda a, b: a and b)
    _bin("or", lambda a, b: a or b)

    # --- unary ---
    _una("not", lambda a: not a)
    _una("neg", lambda a: -a)

    # --- I/O ---
    _una("print", lambda a: print(a, end="") or a)
    _una("println", lambda a: print(a) or a)

    # --- type checks (no '?' suffix — plain WORDs in original tokenizer) ---
    _una("is_null", lambda a: a is None)
    _una("is_bool", lambda a: isinstance(a, bool))
    _una("is_int", lambda a: isinstance(a, int))
    _una("is_float", lambda a: isinstance(a, float))
    _una("is_text", lambda a: isinstance(a, str))
    _una("is_word", lambda a: isinstance(a, str) and a not in ("true", "false", "null"))

    return b


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------

def eval_expr(expr: Any, env: Env) -> Any:
    """Evaluate an AST node in the given environment."""
    # --- Literal values ---
    if isinstance(expr, (int, float, bool, str)) or expr is None:
        return expr

    # --- Symbol lookup ---
    if isinstance(expr, Symbol):
        return env.lookup(expr.name)

    # --- SExpr (S-expression) ---
    if isinstance(expr, SExpr):
        items = expr.items
        if not items:
            return None  # empty list

        first = items[0]

        # Resolve the operator name
        if isinstance(first, Symbol):
            opname = first.name

            # --- Special forms ---

            if opname == "if":
                # (if cond then [else])
                cond = eval_expr(items[1], env)
                if cond:
                    return eval_expr(items[2], env)
                elif len(items) > 3:
                    return eval_expr(items[3], env)
                else:
                    return None

            if opname == "while":
                # (while cond body)
                result: Any = None
                while eval_expr(items[1], env):
                    result = eval_expr(items[2], env)
                return result

            if opname == "do":
                # (do expr1 expr2 ...)  — sequence, return last
                result = None
                for it in items[1:]:
                    result = eval_expr(it, env)
                return result

            if opname == "def":
                # (def name (params) body)
                name_sym = items[1]
                if not isinstance(name_sym, Symbol):
                    raise SyntaxError("def: name must be a symbol")
                params_list = items[2]
                if not isinstance(params_list, SExpr):
                    raise SyntaxError("def: params must be a list")
                param_names = [s.name for s in params_list.items
                               if isinstance(s, Symbol)]
                body = items[3] if len(items) > 3 else None
                fn = Lambda(param_names, body, env)
                env.define(name_sym.name, fn)
                return fn

            if opname == "fun":
                # (fun (params) body)  — anonymous function
                params_list = items[1]
                if not isinstance(params_list, SExpr):
                    raise SyntaxError("fun: params must be a list")
                param_names = [s.name for s in params_list.items
                               if isinstance(s, Symbol)]
                body = items[2] if len(items) > 2 else None
                return Lambda(param_names, body, env)

            if opname == "set":
                # (set name value)
                name_sym = items[1]
                if not isinstance(name_sym, Symbol):
                    raise SyntaxError("set: name must be a symbol")
                val = eval_expr(items[2], env)
                env.set(name_sym.name, val)
                return val

            if opname == "return":
                # (return [value])
                val = eval_expr(items[1], env) if len(items) > 1 else None
                raise Return(val)

        # --- Function application (evaluate operator, then args) ---
        operator = eval_expr(first, env)
        args = [eval_expr(a, env) for a in items[1:]]

        # Built-in
        if isinstance(operator, Builtin):
            if operator.arity == 1:
                if len(args) != 1:
                    raise TypeError(
                        f"{operator.name} expects 1 arg, got {len(args)}")
                return operator.func(args[0])
            elif operator.arity == 2:
                if len(args) != 2:
                    raise TypeError(
                        f"{operator.name} expects 2 args, got {len(args)}")
                return operator.func(args[0], args[1])
            else:
                raise TypeError(f"unsupported builtin arity: {operator.arity}")

        # User-defined lambda
        if isinstance(operator, Lambda):
            return _apply_lambda(operator, args)

        raise TypeError(f"not callable: {operator}")

    raise TypeError(f"cannot evaluate: {expr!r}")


def _apply_lambda(fn: Lambda, args: List[Any]) -> Any:
    """Create a new environment extended from closure, bind params, evaluate body."""
    if len(args) != len(fn.params):
        raise TypeError(
            f"function expects {len(fn.params)} args, got {len(args)}")
    call_env = fn.closure.extend()
    for name, val in zip(fn.params, args):
        call_env.define(name, val)
    try:
        return eval_expr(fn.body, call_env)
    except Return as r:
        return r.value


# ---------------------------------------------------------------------------
# Top-level environment
# ---------------------------------------------------------------------------

def make_global_env() -> Env:
    """Create the global environment with all builtins."""
    env = Env()
    for name, builtin in _make_builtins().items():
        env.define(name, builtin)
    # Also bind true/false/null as literals
    env.define("true", True)
    env.define("false", False)
    env.define("null", None)
    return env


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------

def _main() -> None:
    import sys

    source = open(sys.argv[1]) if len(sys.argv) > 1 else sys.stdin
    text = source.read()

    tokens = tokenize(text)
    if not tokens:
        return

    ast = parse(tokens)
    env = make_global_env()
    try:
        result = eval_expr(ast, env)
        print(result)
    except Exception as e:
        print(f"error: {e}", file=sys.stderr)
        raise


if __name__ == "__main__":
    _main()

File: cpp/eval.cpp
#include "eval.hpp"

#include <cmath>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>

// ---------------------------------------------------------------------------
// Environment
// ---------------------------------------------------------------------------

void Env::define(const std::string &name, std::shared_ptr<Value> value)
{
    bindings[name] = std::move(value);
}

std::shared_ptr<Value> Env::lookup(const std::string &name)
{
    auto it = bindings.find(name);
    if (it != bindings.end())
        return it->second;
    if (parent)
        return parent->lookup(name);
    throw std::runtime_error("undefined: " + name);
}

void Env::set(const std::string &name, std::shared_ptr<Value> value)
{
    auto it = bindings.find(name);
    if (it != bindings.end())
    {
        bindings[name] = std::move(value);
        return;
    }
    if (parent)
    {
        try
        {
            parent->set(name, value);
            return;
        }
        catch (const std::runtime_error &)
        {
        }
    }
    bindings[name] = std::move(value);
}

std::shared_ptr<Env> Env::extend()
{
    return std::make_shared<Env>(std::shared_ptr<Env>(this,
                                                      [](Env *) {})); // non-owning
}

// ---------------------------------------------------------------------------
// Helpers: convert Token to Value atom
// ---------------------------------------------------------------------------

static std::shared_ptr<Value> token_to_atom(const Token &tok)
{
    switch (tok.kind)
    {
    case TokenType::WORD:
        return std::make_shared<Value>(
            std::make_shared<Symbol>(tok.text));
    case TokenType::TEXT:
        return std::make_shared<Value>(tok.text);
    case TokenType::INTEGER:
        return std::make_shared<Value>((int64_t)std::stoll(tok.text));
    case TokenType::FLOAT:
        return std::make_shared<Value>(std::stod(tok.text));
    case TokenType::BOOLEAN:
        return std::make_shared<Value>(tok.text == "true");
    case TokenType::NULL_TYPE:
        return std::make_shared<Value>(nullptr);
    }
    return std::make_shared<Value>(nullptr);
}

// ---------------------------------------------------------------------------
// Parser: flat token list → nested AST
// ---------------------------------------------------------------------------

std::shared_ptr<Value> parse(const std::vector<Token> &tokens)
{
    size_t pos = 0;

    std::function<std::shared_ptr<Value>()> parse_list;
    parse_list = [&]() -> std::shared_ptr<Value>
    {
        std::vector<std::shared_ptr<Value>> items;
        while (pos < tokens.size())
        {
            const Token &tok = tokens[pos];
            if (tok.kind == TokenType::WORD && tok.text == "(")
            {
                pos++; // consume '('
                items.push_back(parse_list());
            }
            else if (tok.kind == TokenType::WORD && tok.text == ")")
            {
                pos++; // consume ')'
                return std::make_shared<Value>(
                    std::make_shared<SExpr>(std::move(items)));
            }
            else
            {
                items.push_back(token_to_atom(tok));
                pos++;
            }
        }
        throw std::runtime_error("unexpected EOF inside list");
    };

    // Top-level: wrap in an implicit do-list if multiple top-level forms
    std::vector<std::shared_ptr<Value>> top;
    while (pos < tokens.size())
    {
        const Token &tok = tokens[pos];
        if (tok.kind == TokenType::WORD && tok.text == "(")
        {
            pos++; // consume '('
            top.push_back(parse_list());
        }
        else if (tok.kind == TokenType::WORD && tok.text == ")")
        {
            throw std::runtime_error("unexpected ')'");
        }
        else
        {
            top.push_back(token_to_atom(tok));
            pos++;
        }
    }

    if (top.size() == 1)
        return top[0];

    // Wrap multiple forms in (do ...)
    auto do_items = std::vector<std::shared_ptr<Value>>();
    do_items.push_back(
        std::make_shared<Value>(std::make_shared<Symbol>("do")));
    for (auto &t : top)
        do_items.push_back(t);
    return std::make_shared<Value>(
        std::make_shared<SExpr>(std::move(do_items)));
}

// ---------------------------------------------------------------------------
// Helpers: promote numbers for arithmetic ops
// ---------------------------------------------------------------------------

static double as_double(const Value &v)
{
    if (v.is_int())
        return (double)v.int_val;
    if (v.is_float())
        return v.float_val;
    throw std::runtime_error("expected number");
}

static std::shared_ptr<Value> num_result(double v)
{
    if (std::floor(v) == v && std::isfinite(v) &&
        v >= (double)INT64_MIN && v <= (double)INT64_MAX)
        return std::make_shared<Value>((int64_t)v);
    return std::make_shared<Value>(v);
}

// ---------------------------------------------------------------------------
// Built-in operators
// ---------------------------------------------------------------------------

static std::unordered_map<std::string, std::shared_ptr<Builtin>>
make_builtins()
{
    std::unordered_map<std::string, std::shared_ptr<Builtin>> b;

    auto make_bin = [&](const std::string &name,
                        std::function<double(double, double)> fn)
    {
        auto builtin = std::make_shared<Builtin>();
        builtin->arity = 2;
        builtin->name = name;
        builtin->binary = [fn](std::shared_ptr<Value> a,
                               std::shared_ptr<Value> b)
        {
            return num_result(fn(as_double(*a), as_double(*b)));
        };
        b[name] = builtin;
    };

    auto make_bin_bool = [&](const std::string &name,
                             std::function<bool(double, double)> fn)
    {
        auto builtin = std::make_shared<Builtin>();
        builtin->arity = 2;
        builtin->name = name;
        builtin->binary = [fn](std::shared_ptr<Value> a,
                               std::shared_ptr<Value> b)
        {
            return std::make_shared<Value>(fn(as_double(*a), as_double(*b)));
        };
        b[name] = builtin;
    };

    auto make_bool_bin = [&](const std::string &name,
                             std::function<bool(bool, bool)> fn)
    {
        auto builtin = std::make_shared<Builtin>();
        builtin->arity = 2;
        builtin->name = name;
        builtin->binary = [fn](std::shared_ptr<Value> a,
                               std::shared_ptr<Value> b)
        {
            return std::make_shared<Value>(
                fn(a->truthy(), b->truthy()));
        };
        b[name] = builtin;
    };

    auto make_una = [&](const std::string &name,
                        std::function<std::shared_ptr<Value>(
                            std::shared_ptr<Value>)>
                            fn)
    {
        auto builtin = std::make_shared<Builtin>();
        builtin->arity = 1;
        builtin->name = name;
        builtin->unary = fn;
        b[name] = builtin;
    };

    // --- arithmetic ---
    make_bin("+", [](double a, double b)
             { return a + b; });
    make_bin("-", [](double a, double b)
             { return a - b; });
    make_bin("*", [](double a, double b)
             { return a * b; });
    make_bin("/", [](double a, double b)
             { return a / b; });
    make_bin("mod", [](double a, double b)
             { return std::fmod(a, b); });

    // --- comparison ---
    make_bin_bool("<", [](double a, double b)
                  { return a < b; });
    make_bin_bool(">", [](double a, double b)
                  { return a > b; });
    make_bin_bool("<=", [](double a, double b)
                  { return a <= b; });
    make_bin_bool(">=", [](double a, double b)
                  { return a >= b; });
    make_bin_bool("=", [](double a, double b)
                  { return a == b; });
    make_bin_bool("!=", [](double a, double b)
                  { return a != b; });

    // --- logic ---
    make_bool_bin("and", [](bool a, bool b)
                  { return a && b; });
    make_bool_bin("or", [](bool a, bool b)
                  { return a || b; });

    // --- unary ---
    make_una("not", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(!a->truthy()); });
    make_una("neg", [](std::shared_ptr<Value> a)
             { return num_result(-as_double(*a)); });

    // --- I/O ---
    make_una("print", [](std::shared_ptr<Value> a)
             {
        if (a->is_int()) std::cout << a->int_val;
        else if (a->is_float()) std::cout << a->float_val;
        else if (a->is_bool()) std::cout << (a->bool_val ? "true" : "false");
        else if (a->is_text()) std::cout << a->text_val;
        else if (a->is_nil()) std::cout << "null";
        else std::cout << "<value>";
        return a; });
    make_una("println", [](std::shared_ptr<Value> a)
             {
        if (a->is_int()) std::cout << a->int_val;
        else if (a->is_float()) std::cout << a->float_val;
        else if (a->is_bool()) std::cout << (a->bool_val ? "true" : "false");
        else if (a->is_text()) std::cout << a->text_val;
        else if (a->is_nil()) std::cout << "null";
        else std::cout << "<value>";
        std::cout << std::endl;
        return a; });

    // --- type checks ---
    make_una("is_null", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_nil()); });
    make_una("is_bool", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_bool()); });
    make_una("is_int", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_int()); });
    make_una("is_float", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_float()); });
    make_una("is_text", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_text()); });
    make_una("is_word", [](std::shared_ptr<Value> a)
             { return std::make_shared<Value>(a->is_text()); });

    return b;
}

// ---------------------------------------------------------------------------
// Evaluator
// ---------------------------------------------------------------------------

static std::shared_ptr<Value>
apply_lambda(std::shared_ptr<Lambda> fn,
             const std::vector<std::shared_ptr<Value>> &args)
{
    if (args.size() != fn->params.size())
        throw std::runtime_error(
            "function expects " + std::to_string(fn->params.size()) +
            " args, got " + std::to_string(args.size()));

    auto call_env = fn->closure->extend();
    for (size_t i = 0; i < fn->params.size(); i++)
        call_env->define(fn->params[i], args[i]);

    try
    {
        return eval_expr(fn->body, call_env);
    }
    catch (const Return &r)
    {
        return r.value();
    }
}

std::shared_ptr<Value> eval_expr(std::shared_ptr<Value> expr,
                                 std::shared_ptr<Env> env)
{
    if (!expr)
        return std::make_shared<Value>(nullptr);

    // --- Literal values ---
    if (expr->is_int() || expr->is_float() || expr->is_bool() ||
        expr->is_text() || expr->is_nil())
        return expr;

    // --- Symbol lookup ---
    if (expr->is_symbol())
        return env->lookup(expr->symbol_val->name);

    // --- SExpr ---
    if (expr->is_sexpr())
    {
        auto &items = expr->sexpr_val->items;
        if (items.empty())
            return std::make_shared<Value>(nullptr);

        auto first = items[0];

        // Resolve the operator name
        if (first->is_symbol())
        {
            const std::string &opname = first->symbol_val->name;

            // --- Special forms ---

            if (opname == "if")
            {
                // (if cond then [else])
                auto cond = eval_expr(items[1], env);
                if (cond->truthy())
                    return eval_expr(items[2], env);
                else if (items.size() > 3)
                    return eval_expr(items[3], env);
                else
                    return std::make_shared<Value>(nullptr);
            }

            if (opname == "while")
            {
                // (while cond body)
                std::shared_ptr<Value> result =
                    std::make_shared<Value>(nullptr);
                while (eval_expr(items[1], env)->truthy())
                    result = eval_expr(items[2], env);
                return result;
            }

            if (opname == "do")
            {
                // (do expr1 expr2 ...)  — sequence, return last
                std::shared_ptr<Value> result =
                    std::make_shared<Value>(nullptr);
                for (size_t i = 1; i < items.size(); i++)
                    result = eval_expr(items[i], env);
                return result;
            }

            if (opname == "def")
            {
                // (def name (params) body)
                auto name_sym = items[1];
                if (!name_sym->is_symbol())
                    throw std::runtime_error(
                        "def: name must be a symbol");

                auto params_list = items[2];
                if (!params_list->is_sexpr())
                    throw std::runtime_error(
                        "def: params must be a list");

                std::vector<std::string> param_names;
                for (auto &p : params_list->sexpr_val->items)
                    if (p->is_symbol())
                        param_names.push_back(p->symbol_val->name);

                auto body =
                    items.size() > 3 ? items[3] : std::make_shared<Value>(nullptr);

                auto fn = std::make_shared<Lambda>(
                    param_names, body, env);

                env->define(name_sym->symbol_val->name,
                            std::make_shared<Value>(fn));
                return std::make_shared<Value>(fn);
            }

            if (opname == "fun")
            {
                // (fun (params) body)  — anonymous function
                auto params_list = items[1];
                if (!params_list->is_sexpr())
                    throw std::runtime_error(
                        "fun: params must be a list");

                std::vector<std::string> param_names;
                for (auto &p : params_list->sexpr_val->items)
                    if (p->is_symbol())
                        param_names.push_back(p->symbol_val->name);

                auto body =
                    items.size() > 2 ? items[2] : std::make_shared<Value>(nullptr);

                return std::make_shared<Value>(
                    std::make_shared<Lambda>(param_names, body, env));
            }

            if (opname == "set")
            {
                // (set name value)
                auto name_sym = items[1];
                if (!name_sym->is_symbol())
                    throw std::runtime_error(
                        "set: name must be a symbol");
                auto val = eval_expr(items[2], env);
                env->set(name_sym->symbol_val->name, val);
                return val;
            }

            if (opname == "return")
            {
                // (return [value])
                auto val = items.size() > 1
                               ? eval_expr(items[1], env)
                               : std::make_shared<Value>(nullptr);
                throw Return(val);
            }
        }

        // --- Function application ---
        auto op = eval_expr(first, env);
        std::vector<std::shared_ptr<Value>> args;
        for (size_t i = 1; i < items.size(); i++)
            args.push_back(eval_expr(items[i], env));

        // Built-in
        if (op->is_builtin())
        {
            auto &builtin = *op->builtin_val;
            if (builtin.arity == 1)
            {
                if (args.size() != 1)
                    throw std::runtime_error(
                        builtin.name + " expects 1 arg, got " +
                        std::to_string(args.size()));
                return builtin.unary(args[0]);
            }
            else if (builtin.arity == 2)
            {
                if (args.size() != 2)
                    throw std::runtime_error(
                        builtin.name + " expects 2 args, got " +
                        std::to_string(args.size()));
                return builtin.binary(args[0], args[1]);
            }
            else
            {
                throw std::runtime_error(
                    "unsupported builtin arity: " +
                    std::to_string(builtin.arity));
            }
        }

        // User-defined lambda
        if (op->is_lambda())
            return apply_lambda(op->lambda_val, args);

        throw std::runtime_error("not callable");
    }

    throw std::runtime_error("cannot evaluate expression");
}

// ---------------------------------------------------------------------------
// Global environment
// ---------------------------------------------------------------------------

std::shared_ptr<Env> make_global_env()
{
    auto env = std::make_shared<Env>();
    for (auto &pair : make_builtins())
        env->define(pair.first,
                    std::make_shared<Value>(pair.second));
    // Also bind true/false/null as literals
    env->define("true", std::make_shared<Value>(true));
    env->define("false", std::make_shared<Value>(false));
    env->define("null", std::make_shared<Value>(nullptr));
    return env;
}